src/Microsoft.Azure.WebJobs.Host/Bindings/BindingBase.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Protocols;
+
+namespace Microsoft.Azure.WebJobs.Host.Bindings
+{
+
+    // Helper class for implementing IBinding with the attribute resolver pattern. 
+    internal abstract class BindingBase<TAttribute> : IBinding
+        where TAttribute : Attribute
+    {
+        protected readonly AttributeCloner<TAttribute> _cloner;
+        private readonly ParameterDescriptor _param;
+
+        public BindingBase(AttributeCloner<TAttribute> cloner, ParameterDescriptor param)
+        {
+            _cloner = cloner;
+            _param = param;
+        }
+
+        public BindingBase(AttributeCloner<TAttribute> cloner, ParameterInfo parameterInfo)
+        {
+            _cloner = cloner;
+            _param = new ParameterDescriptor
+            {
+                Name = parameterInfo.Name,
+                DisplayHints = new ParameterDisplayHints
+                {
+                    Description = "value"
+                }
+            };
+        }
+
+        public bool FromAttribute
+        {
+            get
+            {
+                return true;
+            }
+        }
+
+        protected abstract Task<IValueProvider> BuildAsync(TAttribute attrResolved);
+
+        public Task<IValueProvider> BindAsync(BindingContext context)
+        {
+            var attrResolved = _cloner.ResolveFromBindings(context);
+            return BuildAsync(attrResolved);
+        }
+
+        public Task<IValueProvider> BindAsync(object value, ValueBindingContext context)
+        {
+            var str = value as string;
+            if (str != null)
+            {
+                // Called when we invoke from dashboard. 
+                // str --> attribute --> obj 
+                var resolvedAttr = _cloner.ResolveFromInvokeString(str);
```

Comment:
I don't have a good sense from looking at the code whether this ResolveFromInvokeString pattern works well across the various binding types. I know that for some of the bindings I wrote I've had to do non-trivial conversions where the invoke data doesn't come from the attribute itself. E.g. WebHookTrigger (code [here](https://github.com/Azure/azure-webjobs-sdk-extensions/blob/master/src/WebJobs.Extensions.WebHooks/Bindings/WebHookTriggerBinding.cs#L165)).

How do I perform such operations if I need to?